packages/grafana-data/src/panel/PanelPlugin.ts

```diff
@@ -177,17 +178,41 @@ export class PanelPlugin<
     return this._fieldConfigRegistry;
   }
 
-  get optionEditors(): PanelOptionEditorsRegistry {
-    if (!this._optionEditors) {
-      const builder = new PanelOptionsEditorBuilder<TOptions>();
-      this._optionEditors = builder.getRegistry();
+  private _optionDependencies?: Array<keyof TOptions>;
+  private _lastOptionsItems?: PanelOptionsEditorItem[];
+  private _lastOptions?: TOptions;
 
-      if (this.registerOptionEditors) {
-        this.registerOptionEditors(builder);
+  /**
+   * Returns the editor elements required for the current coptions.
+   */
+  getOptionsEditors(current: TOptions): PanelOptionsEditorItem[] {
```

Comment:
@dprokop -- this is the big change.  Things to note:

1. we never create/use the registry.  For panel properties, we never look things up by id, so :man_shrugging: I think registry was just extra work
2. This adds extra logic/complexity to invalidate the cached items when some keys change.  I'm not sure it is necessary, but worth looking at